presto-main/src/main/java/io/prestosql/operator/LookupJoinOperator.java

```diff
@@ -41,587 +44,679 @@
-    private boolean finishing;
-    private boolean unspilling;
-    private boolean finished;
-    private long joinPosition = -1;
-    private int joinSourcePositions;
-
-    private boolean currentProbePositionProducedRow;
-
-    private final Map<Integer, SavedRow> savedRows = new HashMap<>();
-    @Nullable
-    private ListenableFuture<PartitionedConsumption<Supplier<LookupSource>>> partitionedConsumption;
-    @Nullable
-    private Iterator<Partition<Supplier<LookupSource>>> lookupPartitions;
-    private Optional<Partition<Supplier<LookupSource>>> currentPartition = Optional.empty();
-    private Optional<ListenableFuture<Supplier<LookupSource>>> unspilledLookupSource = Optional.empty();
-    private Iterator<Page> unspilledInputPages = emptyIterator();
-
-    public LookupJoinOperator(
-            OperatorContext operatorContext,
-            List<Type> probeTypes,
+    private final PageBuffer pageBuffer = new PageBuffer();
+    private final WorkProcessor<Page> pages;
+    private final JoinTransformation joinPages;
+
+    public LookupJoinOperator(List<Type> probeTypes,
             List<Type> buildOutputTypes,
             JoinType joinType,
             LookupSourceFactory lookupSourceFactory,
             JoinProbeFactory joinProbeFactory,
             Runnable afterClose,
             OptionalInt lookupJoinsCount,
             HashGenerator hashGenerator,
-            PartitioningSpillerFactory partitioningSpillerFactory)
+            PartitioningSpillerFactory partitioningSpillerFactory,
+            DriverYieldSignal yieldSignal,
+            SpillContext spillContext,
+            MemoryTrackingContext memoryTrackingContext,
+            Optional<WorkProcessor<Page>> sourcePages)
     {
-        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
-        this.probeTypes = ImmutableList.copyOf(requireNonNull(probeTypes, "probeTypes is null"));
-
-        requireNonNull(joinType, "joinType is null");
-        // Cannot use switch case here, because javac will synthesize an inner class and cause IllegalAccessError
-        probeOnOuterSide = joinType == PROBE_OUTER || joinType == FULL_OUTER;
-
-        this.joinProbeFactory = requireNonNull(joinProbeFactory, "joinProbeFactory is null");
-        this.afterClose = requireNonNull(afterClose, "afterClose is null");
-        this.lookupJoinsCount = requireNonNull(lookupJoinsCount, "lookupJoinsCount is null");
-        this.hashGenerator = requireNonNull(hashGenerator, "hashGenerator is null");
-        this.lookupSourceFactory = requireNonNull(lookupSourceFactory, "lookupSourceFactory is null");
-        this.partitioningSpillerFactory = requireNonNull(partitioningSpillerFactory, "partitioningSpillerFactory is null");
-        this.lookupSourceProviderFuture = lookupSourceFactory.createLookupSourceProvider();
-
-        this.statisticsCounter = new JoinStatisticsCounter(joinType);
-        operatorContext.setInfoSupplier(this.statisticsCounter);
```

Comment:
I think we should start by adding support for additional info to `io.prestosql.operator.WorkProcessorOperator`.
Similarly for spill. Otherwise we would get regression feature wise.